libs/sdk-core/src/breez_services.rs

```diff
@@ -1491,6 +1493,34 @@ impl BreezServices {
         Ok(())
     }
 
+    fn detect_network_change(self: &Arc<BreezServices>) {
```

Comment:
with_connection_fallback looks great on node_api. Not sure propagating this to other clients to reconnect is necessary? if we set keep alive to 5 seconds  all other connections will recover more or less at the same time independently.

Reply:
That then depends on the keepalive pings being sent around the same time. When testing this I got an error from `lsp_info`, which was called in `do_sync` after the payment completed. That hanged and timed out. I think reying on those keepalive pings is too unreliable.

Reply:
I agree it won't be at the same time but I think also we can't avoid entirely from transport errors even from the network change detection.
The case you mentioned is indeed interesting because based on my tests if all connections are set to 5 seconds keep alive timeout  and assuming payment takes more than 5 seconds it shouldn't happen because the sync that comes afterward should be already with recovered connections. Are you sure the tests were done with this PR that configures correctly the breez server timeouts: https://github.com/breez/breez-sdk-greenlight/pull/1133

Reply:
No, they weren't done with those breez server timeouts. Greenlight may potentially get different timeouts here: https://github.com/Blockstream/greenlight/pull/548. Also a little calculation is quickly made:
keepalive pings between clients can be 5 seconds apart. Consider the worst case:

keepalive ping from gl-client timed out now, keepalive ping from breez server client times out in 5 seconds. gl-client retries and finishes the call within 2 seconds. There's 3 seconds left to fall into the breez server client timing out.

Reply:
That's right. So basically you are trying to solve the main use cases (payment, invoice) that starts with greenlight and make sure breez server connection that is involved later won't break it. Did I get it right?
I think it is a good goal indeed and I am not against the proposed solution just looking to ensure there is not simpler one.
How about minimizing the breez server keep alive interval and timeout to 1-2 seconds? keep alive is just a small 1 tcp packet which seems very negligible.
No payment/invoice API will be finished in 2 seconds (I think only reconnecting takes more than a second).
If you have a strong opinion that we need the network change detection from greenlight then let's do it.

Reply:
Maybe wrap all (most) grpc calls in with_connection_fallback?

Reply:
Sounds good to me!